### DISCUSSION TITLE
i2s connection between two ESP32

### DISCUSSION THREAD

**Discussion by monojk** [pschatzmann/arduino-audio-tools#149] [Q&A] [answered]
I have a sender ESP32 generating a sine wave (from discussion #26) for testing the connection to a second ESP32:
```
#define SAMPLE_RATE 44100
#define CHANNELS 2
#define BITS_PER_SAMPLE 16

typedef int16_t sound_t; // sound will be represented as int16_t (with 2 bytes)
SineWaveGenerator<sound_t> sineWave(32000); // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<sound_t> sound(sineWave); // Stream generated from sine wave
I2SStream i2sStream;
StreamCopy copier(i2sStream, sound); // copies sound into i2s

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  delay(1000);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  
  // start i2s output
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(TX_MODE);
  config.is_master = true;
  config.pin_data = 22; 
  config.pin_ws   = 15; 
  config.pin_bck  = 14; 
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;
  i2sStream.begin(config);
  Serial.println("I2S started");
  
  // Setup sine wave
  sineWave.begin(CHANNELS, SAMPLE_RATE, N_B4);
}

// Arduino loop - copy sound to i2sStream
void loop() {
  copier.copy();
}
```

The second ESP32 receives a **distorted sine**. What can be the cause? Settings in both ends are the same. 
```
#define SAMPLE_RATE 44100
#define CHANNELS 2
#define BITS_PER_SAMPLE 16

I2SStream i2sStream; // Access I2S as stream
CsvStream<int32_t> csvStream(Serial);
StreamCopy copier(csvStream, i2sStream); // copy i2sStream to csvStream

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  delay(1000);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  // start i2s input
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.is_master = false;
  config.pin_data = 32;
  config.pin_ws   = 15;
  config.pin_bck  = 14;
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;
  i2sStream.begin(config);
  Serial.println("I2S started");

  csvStream.begin(config);
}

void loop() {
  copier.copy();
}
```
![CSV from I2S](https://user-images.githubusercontent.com/25121142/160799597-8d6ac989-190d-4f1a-8433-8a7259089918.jpg)

**Comment by pschatzmann** [comment]
Difficult to say: I am having these kind of effects if the log level is not set to warning, which does not apply in your case.
You can also try to decrease the sample rate and play around with the use_apll setting...

**Reply by monojk** [reply]
lowering the sample rate and using use_apll didn't help. Are there other I2S pins with a better performance?

**Comment by monojk** [comment]
With a sample rate of 16000 there is still distortion, yet not so often but still too much. I set the use_apll on the receiver side. The sender is the master, the receiver is master=false.  
![CSV from I2S](https://user-images.githubusercontent.com/25121142/161044384-aaca016f-3b0f-4f60-811b-d176e4c13a9b.jpg)

**Comment by softhack007** [comment]
The plot looks like one of the two ESP32 is dropping samples, most likely cause by buffer overrun. Also your observation that lowering the sample rate leads to less distortion could mean that one of your ESP32 does not have enough buffer space to store samples before you "collect" them in your sketch.

I am not an expert on Arduino-audio-tools, but generally speaking it might help to increase the i2s buffer size, or increase number of buffers or adjust both parameters. Did you try this already?

**Comment by yujonglee** [comment]
@pschatzmann

I tried with config below, but still experiencing distortion. Can you give me any suggestions please?
```
#define SAMPLE_RATE 16000
config.use_apll = true;
config.buffer_size = 1024;
config.buffer_count = 8;
```

Full code.

```
#include "AudioTools.h"

#define SAMPLE_RATE 16000
#define CHANNELS 2
#define BITS_PER_SAMPLE 16

typedef int16_t sound_t; // sound will be represented as int16_t (with 2 bytes)
SineWaveGenerator<sound_t> sineWave(32000); // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<sound_t> sound(sineWave); // Stream generated from sine wave
I2SStream i2sStream;
StreamCopy copier(i2sStream, sound); // copies sound into i2s

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  delay(1000);
  // AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  
  // start i2s output
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(TX_MODE);
  config.is_master = true;
  config.pin_data = 22; 
  config.pin_ws   = 15; 
  config.pin_bck  = 14; 
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;

  config.use_apll = true;
  config.buffer_size = 1024;
  config.buffer_count = 8;

  i2sStream.begin(config);
  Serial.println("I2S started");
  
  // Setup sine wave
  sineWave.begin(CHANNELS, SAMPLE_RATE, N_B4);
}

// Arduino loop - copy sound to i2sStream
void loop() {
  copier.copy();
}
```

```
#include "AudioTools.h"

#define SAMPLE_RATE 16000
#define CHANNELS 2
#define BITS_PER_SAMPLE 16

I2SStream i2sStream; // Access I2S as stream
CsvStream<int32_t> csvStream(Serial);
StreamCopy copier(csvStream, i2sStream); // copy i2sStream to csvStream

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  delay(1000);
  // AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  // start i2s input
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.is_master = false;
  config.pin_data = 32;
  config.pin_ws   = 15;
  config.pin_bck  = 14;
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;

  config.use_apll = true;
  config.buffer_size = 1024;
  config.buffer_count = 8;

  i2sStream.begin(config);
  Serial.println("I2S started");

  csvStream.begin(config);
}

void loop() {
  copier.copy();
}
```

**Reply by yujonglee** [reply]
<img width="817" alt="스크린샷 2023-04-25 오후 12 20 46" src="https://user-images.githubusercontent.com/61503739/234166434-323ecbe1-1718-495e-8323-929c2d7f48d0.png">

**Comment by pschatzmann** [comment]
I never tried this myself, so I don't know if this helps:
- keep the cables as short as possible
- try to have the receiver drive the process and define it as master and the sender as slave

**Reply by yujonglee** [reply]
@pschatzmann 
Thanks for the recommendations. But it didn't solved my issue.

Some more observations:

1. Distortion occurs at regular time intervals.
2. The sample doesn't change or drop, but rather seems to cut off in certain places. In other words, the period of the Sine wave changes at that moment, and then it becomes constant again. (image above in my last comment)
3. When the bck Pin is removed, a smooth sine wave appears for a few seconds without any problem, and then stops.

I don't have much audio programming knowledge, so this could be completely bizarre, but I wonder if the problem is caused by the clock synchronization of the two devices at regular intervals.

The two devices are connected by a 5 cm cable.

I'd love to hear your thoughts on this situation.
Thanks for your time.

**Reply by pschatzmann** [reply]
I am wondering that this is working at all! If you looked at your posted source code you messed up badly with the bits_per_sample and the corresponding data types: bits_per_sample of 16 needs to be int16_t for CsvStream!

**Comment by jamesbaber** [comment]
I have this same setup at the moment as I thought it would be perfect to test the library out before buying I2S peripherals for my project. I'm using streams-generator-i2s.ino and streams-i2s-serial_16bit.ino

I seem to be getting the same issues, although they don't seem to occur at regular intervals.
It is affected by changing the sample frequency and channel count as you've discussed. Increasing the buffer size and enabling apll doesn't seem to help. 

On my two identical, brand new ESP32s, I am using the default I2S pin mappings.
I have pin 14 connected to 14.
15 to 15.
22 to 32.
GND to GND.
And they are both powered over USB from the same source.

Did you find a solution, @yujonglee, @monojk?

Thanks!

**Comment by pschatzmann** [comment, marked as answer]
I came to the conclusion that the best way to transmit audio via wires, is to use the serial protocol together with a codec.

**Reply by jamesbaber** [reply]
Thanks for your reply!
Are you saying it's specifically the I2S TX that's problematic? Or is it a general stability issue when ESP32s handle both ends?

**Reply by pschatzmann** [reply]
ESP32 can handle both ends: sending audio to an external DAC and receiving from an external ADC is reliable.

Just the transfer between two microcontrollers seems to have issues

**Comment by SamObde** [comment]
I have been troubleshooting this ESP32 to ESP32 via I2S problem a week now.  I finally got a logic analyzer and think I have arrived at part of the problem.  The BCLK seems to be sorta unstable.  In this first picture you can see the clock is mostly at 1.42mhz...

![image](https://github.com/user-attachments/assets/f2585d90-3d5b-4fe9-b114-1ee6e5fbc18b)

But in this second picture you can see every once in a while there is a 1.25mhz blip.  It happens pretty frequently.  I am using APLL.  I upgraded to SDK 5.3.2 which helped quite a bit but its still not stable.  I have some new ESP32 devkits coming directly from Espressif to see if maybe the older revision 0 and 1 or so chips have problems. 

![image](https://github.com/user-attachments/assets/21b39a19-917a-4803-9cb7-2d84c266e97a)

I also see the 44.1mhz rate is not quite right ... see below - that one shows 44.05mhz and the next one shows 44.24mhz.  I am not sure if this is because the BCLK is off or something else. 

![image](https://github.com/user-attachments/assets/b9bf5718-5aae-4f92-94c0-934373750c51)

![image](https://github.com/user-attachments/assets/3c6c1a19-88be-4637-8378-0862c7d22ce9)

I am going to try getting a better external clock to use to see if it helps.  So far I have improved the audio 98% from what I had by just using the new SDK.  I still have little crackles.  I also tried an ESP32-S3 just simply outputting a clock signal - same thing.  I can't seem to get an exact clock out of these things.

**Reply by pschatzmann** [reply]
A work around for the crackles would be to apply a MedianFilter

**Reply by SamObde** [reply]
I will try that - thanks! Do you think the audio imperfectness is from the clock not being perfect?  I have a clock generator coming to test with as well.

**Reply by pschatzmann** [reply]
My gut feeling is that this is coming from slight clock differences between the differernt  devices.
It does not help that the masterclock can only be used as output but not as input....

**Reply by SamObde** [reply]
I think in 5.4 sdk it shows that it can be set as input?  I2S_CLK_SRC_EXTERNAL is defined for the gpio pin now I believe.

[https://docs.espressif.com/projects/esp-idf/en/v5.4/esp32/api-reference/peripherals/i2s.html#_CPPv421i2s_std_gpio_config_t](https://docs.espressif.com/projects/esp-idf/en/v5.4/esp32/api-reference/peripherals/i2s.html#_CPPv421i2s_std_gpio_config_t)

**Reply by pschatzmann** [reply]
Strange: this value does not exist: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/clk_tree.html#_CPPv424soc_periph_i2s_clk_src_t

**Reply by SamObde** [reply]
I think the documentation just hasn't caught up - I believe I see it in the actual code.  It is only supported for SOC version 2 (S3 etc?).

**Reply by pschatzmann** [reply]
I committed a change to support I2S_CLK_SRC_EXTERNAL

**Reply by SamObde** [reply]
You rock!

**Reply by pschatzmann** [reply]
Bad news: I could not get it to work. Whatever I tried, I was getting the following errors

E (4017) i2s_std: i2s_std_calculate_clock(62): sample rate or mclk_multiple is too large for the current clock source
E (4018) i2s_std: i2s_std_set_clock(79): clock calculate failed

However, I noticed that I did not handle the use_apll in the V1 implementation. I committed a correction for this now...

**Reply by SamObde** [reply]
Setting it as I2S_MCLK_MULTIPLE_256?  I guess bad news and good news.  Good news is if use_apll wasn't actually working then all my problems might be fixed : )  I'll try this today.  Thanks!

**Reply by SamObde** [reply]
I have added this into my sketch to try to see if use_apll is active:

```
void checkAPLLStatus(i2s_port_t i2s_port = I2S_NUM_0) {
  
    uint32_t reg_val = REG_READ(I2S_CLKM_CONF_REG(i2s_port));

    Serial.printf("I2S%d Clock Register: 0x%08X\n", i2s_port, reg_val);

    if (reg_val & (1 << 22)) {  // Bit 22 = APLL enable
        Serial.println("✅ APLL is ENABLED for I2S.");
    } else {
        Serial.println("❌ APLL is NOT enabled.");
    }
}
```

I get this output:

```
<!-- STARTUP of GroupTooth MASTER, Firmware: 2025-DEV1 Compile Time: Jan 31 2025 07:44:41 --->
<!-- SDK firmware is v5.3.2-282-gcfea4f7c98-dirty --->
Starting I2S...
Setup Complete.
I2S0 Clock Register: 0x0017450E
❌ APLL is NOT enabled.
```

Here is how I init the i2s driver thru audio-tools:

```
  // start i2s input with default configuration
  Serial.println("Starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.is_master = true;
  cfg.port_no = 0;
  cfg.pin_bck = 27;
  cfg.pin_ws = 26;
  cfg.pin_data = 25;
  cfg.pin_data_rx = 25;
  cfg.use_apll = 1;
  cfg.auto_clear = true;
  i2s.begin(cfg);
```
I am using this branch of the code - I don't 100% know what I am doing but I think I got your latest commit:

https://github.com/pschatzmann/arduino-audio-tools.git#f24d58f16f06a0c02ef60b2bb3f56f3073b3d0f2

Am I doing this wrong?  Thanks!

**Reply by pschatzmann** [reply]
You will need to define the master clock pin,

Set the log level to info!

Dependent on the variant you should see
            LOGI("clk_src is I2S_CLK_SRC_APLL");
or 
            LOGI("clk_src is I2S_CLK_SRC_PLL_160M");

What esp32 variant are you using ?
